### ISSUE TITLE
The current type checker fails to detect a type mismatch involving type interface

### ISSUE THREAD

**Issue by JacksonGL** [google/closure-compiler#980]
The last two assignments in function ```f``` should trigger warnings in the current type system.
```javascript
/**
 * @interface
 * @template KEY1, VALUE1
 */
var IObject = function() {};

/**
 * @interface
 * @extends {IObject<number, VALUE2>}
 * @template VALUE2
 */
var IArrayLike = function() {};

/**
 * @constructor
 * @implements {IArrayLike<number>}
 */
function Int8Array2() {}

function f() {
  /** @type {Int8Array2} */
  var x = new Int8Array2();

  /** @type {IArrayLike<string>} */
  var y;
  y = x;  // should have a warning here

  /** @type {IObject<number, string>} */
  var z;
  z = x;  // should have a warning here
}
```

**Comment by dimvar** [comment]
FYI, this works correctly in the new type inference. You'll have to fix it in OTI in order to land IObject and IArrayLike.

**Comment by JacksonGL** [comment]
I see, I am working on it.

**Comment by JacksonGL** [comment]
I think I find the reason. It turns out that in the old type system, ```Int8Array2``` is a function literal and therefore hoisted. The hoisted function declaration is processed before ```IObject``` and ```IArrayLike```. Consequently, its template type map data is somehow incomplete. This leads to ```KEY_MISMATCH``` (for template type match) when doing the type equivalence check (in the type checking stage). Therefore the types are considered equivalent under ```INVARIANT``` match. Therefore the type check missed. Not sure if this should be considered as a bug. 

The following example works fine:
```javascript
/**
* @interface
* @template KEY1, VALUE1
*/
function IObject() {}; // hoisted

/**
* @interface 
* @extends {IObject<number, VALUE2>}
* @template VALUE2
*/
function IArrayLike() {}; // hoisted

/**
* @constructor 
* @implements {IArrayLike<number>}
*/
var Int8Array2 = function() {}; // not hoisted

function f() {
  /** @type {Int8Array2} */
  var x = new Int8Array2();

  /** @type {IArrayLike<string>} */
  var y;
  y = x;  // warning generated

  /** @type {IObject<number, string>} */
  var z;
  z = x;  // warning generated
}
```

While in the following case, type checking partially works because of hoistering only one super class.
```javascript
/**
* @interface
* @template KEY1, VALUE1
*/
var IObject = function() {};

/**
* @interface 
* @extends {IObject<number, VALUE2>}
* @template VALUE2
*/
function IArrayLike() {}; // hoisted, processed before IObject, template information from IObject is incomplete

/**
* @constructor 
* @implements {IArrayLike<number>}
*/
var Int8Array2 = function() {}

function f() {
  var x = new Int8Array2();

  /** @type {IArrayLike<string>} */
  var y;
  y = x; // warning generated

  /** @type {IObject<boolean, number>} */
  var z;
  z = x; // no warning, as KEY1 (from IObject) is incomplete
}

```

**Comment by dimvar** [comment]
This is definitely a bug. You should be able to declare functions in externs without using assignment, and without changes in the typing results. Eg, functions are declared like that in es3.js.

https://github.com/google/closure-compiler/blob/master/externs/es3.js
